x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/shards/FrozenShardsDeciderService.java

```diff
@@ -0,0 +1,138 @@
+import org.elasticsearch.xpack.cluster.routing.allocation.DataTierAllocationDecider;
+import org.elasticsearch.xpack.core.DataTier;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.StreamSupport;
+
+/**
+ * This decider enforces that on a 64GB memory node (31GB heap) we can max have 2000 shards. We arrive at 2000 because our current limit is
+ * 1000 but frozen tier uses the "frozen engine", which is much more efficient. We scale the total tier memory accordingly.
+ *
+ * The decider relies on frozen tier being used exclusively for frozen shards.
+ */
+public class FrozenShardsDeciderService implements AutoscalingDeciderService {
+    public static final String NAME = "frozen_shards";
+    private static final ByteSizeValue MAX_MEMORY = ByteSizeValue.ofGb(64);
+    static final ByteSizeValue DEFAULT_MEMORY_PER_SHARD = ByteSizeValue.ofBytes(MAX_MEMORY.getBytes() / 2000);
+    public static final Setting<ByteSizeValue> MEMORY_PER_SHARD = Setting.byteSizeSetting(
+        "memory_per_shard",
+        (dummy) -> DEFAULT_MEMORY_PER_SHARD.getStringRep()
+    );
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public AutoscalingDeciderResult scale(Settings configuration, AutoscalingDeciderContext context) {
+        // we assume that nodes do not grow beyond 64GB here.
+        long shards = countFrozenShards(context.state().metadata());
+        long memory = shards * MEMORY_PER_SHARD.get(configuration).getBytes();
+        return new AutoscalingDeciderResult(AutoscalingCapacity.builder().total(null, memory).build(), new FrozenShardsReason(shards));
+    }
+
+    static long countFrozenShards(Metadata metadata) {
+        return StreamSupport.stream(metadata.spliterator(), false)
+            .filter(imd -> isFrozenIndex(imd.getSettings()))
+            .mapToLong(IndexMetadata::getTotalNumberOfShards)
+            .sum();
+    }
+
+    static boolean isFrozenIndex(Settings indexSettings) {
+        String tierPreference = DataTierAllocationDecider.INDEX_ROUTING_PREFER_SETTING.get(indexSettings);
+        String[] preferredTiers = DataTierAllocationDecider.parseTierList(tierPreference);
+        if (preferredTiers.length >= 1 && preferredTiers[0].equals(DataTier.DATA_FROZEN)) {
+            assert preferredTiers.length == 1 : "frozen tier preference must be frozen only";
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public List<Setting<?>> deciderSettings() {
+        return List.of(MEMORY_PER_SHARD);
+    }
+
+    @Override
+    public List<DiscoveryNodeRole> roles() {
+        return List.of(DiscoveryNodeRole.DATA_FROZEN_NODE_ROLE);
+    }
+
+    public static class FrozenShardsReason implements AutoscalingDeciderResult.Reason {
+        private final long shards;
+
+        public FrozenShardsReason(long shards) {
+            this.shards = shards;
```

Comment:
Should we validate this is non-negative, and then use `(read|write)Vlong()` in serialization? It's not a huge deal, just curious whether we expect this to ever be negative.

Reply:
++, done in 8f975a065beb91f467b37268119516d5eda6092e